### PR TITLE
first spike of the GitOps mode for jx install

### PR DESCRIPTION
**Pull request by jstrachan**
which creates a development Environment source code tree which can be applied via `jx step env apply` in a CI/CD pipeline

### CODE REVIEW

File: pkg/helm/helm_helpers.go [hunk 1]
```diff
@@ -263,3 +271,53 @@ func AppendMyValues(valueFiles []string) ([]string, error) {
...
+	}
+	err = ioutil.WriteFile(outFile, []byte(text), util.DefaultWritePermissions)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to save combined helm values YAML file %s", outFile)
+	}
+	return nil
+}
+
+// CombineMapTrees recursively copies all the values from the input map into the destination map preserving any missing entries in the destination
+func CombineMapTrees(destination map[string]interface{}, input map[string]interface{}) {
```
Comment (ccojocar) [inline pkg/helm/helm_helpers.go]: I would move this into `pkg/utils/maps.go` that can be reused also in other parts of the oded.

File: pkg/jx/cmd/install.go [hunk 2]
```diff
@@ -920,58 +1035,77 @@ func (options *InstallOptions) Run() error {
...
+		}
 
-	for _, ac := range addonConfig.Addons {
-		if ac.Enabled {
-			err = options.installAddon(ac.Name)
-			if err != nil {
-				return fmt.Errorf("failed to install addon %s: %s", ac.Name, err)
+		for _, ac := range addonConfig.Addons {
+			if ac.Enabled {
+				err = options.installAddon(ac.Name)
```
Comment (ccojocar) [inline pkg/jx/cmd/install.go]: Do we also store the addon custom values somewhere in the gitops config?

**Comment by rawlingsj** [comment]
I tried out the PR last night with prow and for some reason everything was installed in the `default` namespace.  There was no `jx` namespace created.  This is from the install:

```
? Enter the new repository name:  environment-forgerplump-staging


Creating repository cb-kubecd/environment-forgerplump-staging
Creating Git repository cb-kubecd/environment-forgerplump-staging
Pushed Git repository to https://github.com/cb-kubecd/environment-forgerplump-staging

? Git branch for the Environment source code: master
Created environment staging
error creating cluster failed to create staging environment in namespace jx: failed to add repo cb-kubecd/environment-forgerplump-staging to Prow config in namespace jx-staging: namespaces "jx" not found
error: failed to create staging environment in namespace jx: failed to add repo cb-kubecd/environment-forgerplump-staging to Prow config in namespace jx-staging: namespaces "jx" not found
➜  jx git:(gitops) ✗ kc get ns
NAME          STATUS   AGE
default       Active   22m
kube-public   Active   22m
kube-system   Active   22m
➜  jx git:(gitops) ✗ kc get pods
NAME                                READY   STATUS    RESTARTS   AGE
build-controller-5769957899-6x9cp   1/1     Running   0          5m11s
deck-6d7985c6-lrscx                 1/1     Running   0          6m30s
deck-6d7985c6-z45qv                 1/1     Running   0          6m30s
hook-c75964c6f-pm7fd                1/1     Running   0          6m30s
hook-c75964c6f-ql7wx                1/1     Running   0          6m30s
horologium-559f4558b7-zt64c         1/1     Running   0          6m30s
plank-7f968b8cc7-qf4d7              1/1     Running   0          6m29s
prow-build-dc4cc9d8c-rpskq          1/1     Running   0          6m30s
sinker-c4c5b77d8-wrktq              1/1     Running   0          6m29s
tide-5fcc5f88c7-gjwqx               1/1     Running   0          6m29s
```

Command used:
```
jx create cluster gke --prow  --skip-login=true --machine-type n1-standard-2 --min-num-nodes=3 --max-num-nodes=3 --default-admin-password=foo --zone=europe-west1-b --labels test=install
```

**Pull request merged**